moonraker/plugins/power.py

```diff
@@ -6,94 +6,84 @@
 
 import logging
 import os
+import asyncio
+import json
+import struct
+import socket
+import gpiod
 from tornado.ioloop import IOLoop
+from tornado.iostream import IOStream
 from tornado import gen
 
 class PrinterPower:
     def __init__(self, config):
         self.server = config.get_server()
         self.server.register_endpoint(
-            "/machine/gpio_power/devices", ['GET'],
+            "/machine/device_power/devices", ['GET'],
             self._handle_list_devices)
         self.server.register_endpoint(
-            "/machine/gpio_power/status", ['GET'],
+            "/machine/device_power/status", ['GET'],
             self._handle_power_request)
         self.server.register_endpoint(
-            "/machine/gpio_power/on", ['POST'],
+            "/machine/device_power/on", ['POST'],
             self._handle_power_request)
         self.server.register_endpoint(
-            "/machine/gpio_power/off", ['POST'],
+            "/machine/device_power/off", ['POST'],
             self._handle_power_request)
         self.server.register_remote_method(
             "set_device_power", self.set_device_power)
 
+        self.chip_factory = GpioChipFactory()
         self.current_dev = None
         self.devices = {}
-        dev_names = config.get('devices')
-        dev_names = [d.strip() for d in dev_names.split(',') if d.strip()]
-        logging.info("Power plugin loading devices: " + str(dev_names))
-        devices = {}
-        for dev in dev_names:
-            devices[dev] = GpioDevice(dev, config)
-        ioloop = IOLoop.current()
-        ioloop.spawn_callback(self.initialize_devices, devices)
+        prefix_sections = config.get_prefix_sections("power")
+        logging.info(f"Power plugin loading devices: {prefix_sections}")
+        for section in prefix_sections:
+            cfg = config[section]
+            dev_type = cfg.get("type")
+            if dev_type == "gpio":
+                dev = GpioDevice(cfg, self.chip_factory)
+            elif dev_type == "tplink_smartplug":
+                dev = TPLinkSmartPlug(cfg)
+            else:
```

Comment:
I'd suggest
```
elif dev_type is None:
  continue
```

or  few lines above:

```
if "type" not in cfg:
  continue
```

I need to load power plugin, to later add my devices, but I don't want to add dummy gpio device, just so plagin will initialize and won't raise an error.

Reply:
I don't believe that will be necessary.  Presumably you will be calling `server.load_plugin('power', config)` when you init your module.  If you aren't adding a [power xxx] section to your config, then `config.get_prefix_sections()` will evaluate to an empty list and the for loop wont be entered.

Reply:
ok, that worked